src/utils/dom/awaitSelector.js

```diff
@@ -0,0 +1,132 @@
+/*
+Copyright 2019 Adobe. All rights reserved.
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+
+import isFunction from "../isFunction";
+import isNonEmptyArray from "../isNonEmptyArray";
+import Promise from "../Promise";
+import selectNodes from "./selectNodes";
+
+const MUTATION_OBSERVER = "MutationObserver";
+const RAF = "requestAnimationFrame";
+const MUTATION_OBSERVER_CONFIG = { childList: true, subtree: true };
+const VISIBILITY_STATE = "visibilityState";
+const VISIBLE = "visible";
+const DELAY = 100;
+const MAX_POLLING_TIMEOUT = 5000;
+
+function createError(selector) {
+  return new Error(`Could not find: ${selector}`);
+}
+
+function createPromise(executor) {
+  return new Promise(executor);
+}
+
+export function canUseMutationObserver(win) {
+  return isFunction(win[MUTATION_OBSERVER]);
+}
+
+export function awaitUsingMutationObserver(
+  win,
+  doc,
+  selector,
+  timeout,
+  selectFunc
+) {
+  return createPromise((resolve, reject) => {
+    const mutationObserver = new win[MUTATION_OBSERVER](() => {
+      const nodes = selectFunc(selector);
+
+      if (isNonEmptyArray(nodes)) {
+        mutationObserver.disconnect();
+        resolve(nodes);
+      }
+    });
+
+    setTimeout(() => {
+      mutationObserver.disconnect();
+      reject(createError(selector));
+    }, timeout);
+
+    mutationObserver.observe(doc, MUTATION_OBSERVER_CONFIG);
+  });
+}
+
+export function canUseRequestAnimationFrame(doc) {
+  return doc[VISIBILITY_STATE] === VISIBLE;
+}
+
+export function awaitUsingRequestAnimation(win, selector, timeout, selectFunc) {
+  return createPromise((resolve, reject) => {
+    function execute() {
+      const nodes = selectFunc(selector);
+
+      if (isNonEmptyArray(nodes)) {
+        resolve(nodes);
+        return;
+      }
+
+      win[RAF](execute);
```

Comment:
Selecting elements on every animation frame without any delay seems like it could be expensive and impact the user experience since it would try to run 60 times a second. I see it being recommended online without much mention of degraded performance though, so maybe I'm just not understand something fundamental or maybe element selection is just always fast enough to not be noticeable? There's some mention of combining `requestAnimationFrame` with a delay, like https://gist.github.com/chrisjhoughton/7890303#gistcomment-2334190. Anyway, I don't know that anything needs to be done here; I just thought I'd express my initial reaction.

Reply:
@Aaronius yeah, initially I was thinking that using RAF might be too aggressive, but we are using a similar approach in Adobe Target and it works great. Also this is mostly a fallback mechanism for IE 10, ideally we should use MutationObserver all the time.